datafusion/physical-optimizer/src/pruning.rs

```diff
@@ -527,6 +529,7 @@ impl PruningPredicate {
     /// See the struct level documentation on [`PruningPredicate`] for more
     /// details.
     pub fn try_new(expr: Arc<dyn PhysicalExpr>, schema: SchemaRef) -> Result<Self> {
+        let expr = snasphot_physical_expr(expr)?;
```

Comment:
This was the other piece of the trick: we need to snapshot here so that stats based pruning works.

Because we do it in `PruningPredicate` even a custom `DataSourceExec`, `TableProvider` or anything else that uses `PruningPredicate` will benefit from dynamic filter pushdown (as long as it's not building the predicate too early).

Reply:
This is good context to add as a comment Ithink 

```suggestion
        // Get a (simpler) snapshot of the physical expr here to use with `PruningPredicate`
        // which does not handle dynamic exprs  in general
        let expr = snasphot_physical_expr(expr)?;
```